libs/runtime/ebpf_object.c

```diff
@@ -148,98 +128,41 @@ ebpf_object_update_reference_history(void* object, bool acquire, uint32_t file_i
-    state = ebpf_lock_lock(&_ebpf_object_tracking_list_lock);
-    for (new_index = 1; new_index < EBPF_COUNT_OF(_ebpf_id_table); new_index++) {
-        if (_ebpf_id_table[new_index].object == NULL && _ebpf_id_table[new_index].reference_count == 0) {
-            break;
-        }
-    }
-    if (new_index == EBPF_COUNT_OF(_ebpf_id_table)) {
-        return_value = EBPF_NO_MEMORY;
-        goto Done;
-    }
-
-    // Generate a new ID.
-    _ebpf_id_table[new_index].counter++;
-    _ebpf_id_table[new_index].reference_count = 1;
-    _ebpf_id_table[new_index].object = object;
-    _update_reference_history(&_ebpf_id_table[new_index], EBPF_OBJECT_CREATE, file_id, line);
-    object->id = _get_id_from_index(new_index);
-
-    return_value = EBPF_SUCCESS;
-
-Done:
-    ebpf_lock_unlock(&_ebpf_object_tracking_list_lock, state);
-
-    return return_value;
-}
-
-_Requires_lock_held_(&_ebpf_object_tracking_list_lock) static void _ebpf_object_tracking_list_remove(
-    _In_ const ebpf_core_object_t* object, ebpf_file_id_t file_id, uint32_t line)
-{
-    uint32_t index;
-    ebpf_result_t return_value = _get_index_from_id(object->id, &index);
+    ebpf_id_entry_t* entry = NULL;
+    ebpf_result_t return_value = ebpf_hash_table_find(_ebpf_id_table, (const uint8_t*)&object->id, (uint8_t**)&entry);
     ebpf_assert(return_value == EBPF_SUCCESS);
 
-    // In a release build, ebpf_assert is a no-op so we need to avoid an unreferenced variable warning.
-    UNREFERENCED_PARAMETER(return_value);
+    ebpf_assert(entry->object == object);
+    entry->object = NULL;
 
-    // Under lock, so un-protected access is ok.
-    _ebpf_id_table[index].reference_count--;
-    _update_reference_history(&_ebpf_id_table[index], EBPF_OBJECT_RELEASE, file_id, line);
-    ebpf_assert(_ebpf_id_table[index].reference_count >= 0);
-    _ebpf_id_table[index].object = NULL;
+    ebpf_object_release_id_reference(object->id, object->type, file_id, line);
 }
 
-void
+ebpf_result_t
 ebpf_object_tracking_initiate()
 {
-    ebpf_lock_create(&_ebpf_object_tracking_list_lock);
-    memset(_ebpf_id_table, 0, sizeof(_ebpf_id_table));
+    _ebpf_next_id = 1;
+    ebpf_hash_table_creation_options_t options = {
+        .key_size = sizeof(ebpf_id_t),
+        .value_size = sizeof(ebpf_id_entry_t),
+        .max_entries = 1024,
+    };
+
     memset(_ebpf_object_reference_history, 0, sizeof(_ebpf_object_reference_history));
     _ebpf_object_reference_history_index = 0;
+
+    return ebpf_hash_table_create(&_ebpf_id_table, &options);
 }
 
 void
 ebpf_object_tracking_terminate()
 {
-    for (int index = 0; index < EBPF_COUNT_OF(_ebpf_id_table); index++) {
-        ebpf_assert(_ebpf_id_table[index].object == NULL || ebpf_fault_injection_is_enabled());
-        ebpf_assert(_ebpf_id_table[index].reference_count == 0 || ebpf_fault_injection_is_enabled());
-    }
+    ebpf_assert(!_ebpf_id_table || ebpf_hash_table_key_count(_ebpf_id_table) == 0 || ebpf_fault_injection_is_enabled());
```

Comment:
Wouldn't invoking this call without a prior invocation of the `ebpf_object_tracking_initiate` spuriously trigger this assert, say if there a failure in the initial startup and we don't call `ebpf_object_tracking_initiate` at all?

(As is, this call is called unconditionally from `ebpf_core_terminate`)

Reply:
I don't follow your logic. The assert is satisfied if one of the following is met:
1) _ebpf_id_table is NULL -> ebpf_object_tracking_initiate was not called or failed.
2) _ebpf_id_table  is not null and empty.
3) ebpf_fault_injection_is_enabled

Under what condition will the assert fire incorrectly?

Reply:
Nm. I was mis-reading the assert.